spirv_msl.cpp

```diff
@@ -393,6 +428,15 @@ uint32_t CompilerMSL::add_interface_block(StorageClass storage)
 	MemberSorter memberSorter(ib_type, meta[ib_type_id], sort_aspect);
 	memberSorter.sort();
 
+	// Sort input or output variables alphabetical
+	auto &execution = get_entry_point();
+	if ((execution.model == ExecutionModelFragment && storage == StorageClassInput) ||
+	    (execution.model == ExecutionModelVertex && storage == StorageClassOutput))
+	{
+		MemberSorter memberSorter(ib_type, meta[ib_type.self], MemberSorter::Alphabetical);
+		memberSorter.sort();
+	}
```

Comment:
The value of _`N`_ in the `[[user(locnN)]]` qualifier is determined by the `get_ordered_member_location()` function from the `DecorationLocation` decoration as specified in the incoming SPIR-V.

If no `DecorationLocation` decoration has been specified, then the `get_ordered_member_location()` function uses the index of the member in the structure.

Sorting alphabetically seems like quite a fragile design that depends on very deliberate agreement between the exact order of the members of the vertex output structure and the vertex input structure.

In the absence of `DecorationLocation` decorations, the alphabetic sorting will only work in the special case where the member indexes align between the vertex output and fragment input structures. In your example, fortunately, the builtins are sorted so that they appear at the end of the structure, otherwise the `gl_Position` member would interfere with the member index of the `texCoord` member, because a strict alphabetic sorting would result in:

```
struct painter_text_vert_main_out
{
    float4 fragmentColor [[user(locn0)]];
    float4 gl_Position [[position]];
    float2 texCoord [[user(locn2)]];
};
```
Notice that `texCoord ` is now set to `[[user(locn2)]]`, because it now sits at member index 2 in this structure.

So...in the general case, any variation between the content of the two structures ahead of the member you are trying to assign a `[[user(locnN)]]` qualifier to will cause mismatches. This might occur, for example, if the vertex output structure outputs more than the fragment input structure uses...such as when a general vertex shader is used with several different fragment shaders.

Is there some reason why your SPIR-V is not specifying a `DecorationLocation` decoration for these inputs and outputs?

In the absence of an explicit `DecorationLocation` decoration, a  more reliable method of matching might be to let Metal perform the matching using names.

We could do this by modifying the `get_ordered_member_location()` function to return an indication that a location has not been defined (maybe return `-1`), and then have the `member_attribute_qualifier()` function return an empty string for that member. Metal will then revert to matching by member names between the two structures.